src/BugManager.php

```diff
@@ -53,9 +64,88 @@ public function submit(Request $req) {
 			return null;
 		}
 
-		$info = yield $this->submit->execute(["title" => $title, "data" => $data]);
+		$add = [];
+		$promises = [];
+		foreach ($this->fields as $field => $info) {
+			$attrs = array_filter($body->getArray($field), function ($x) { return $x != ""; });
+
+			if (empty($info["multi"]) && count($attrs) > 1) {
+				return null;
+			}
+
+			if (count($attrs) < ($info["required"] ?? 0)) {
+				return null;
+			}
+
+			foreach ($attrs as $attr) {
+				switch ($info["type"]) {
+					case self::USER:
+						if (filter_var($attr, FILTER_VALIDATE_INT)) {
+							$promises[] = \Amp\pipe($this->db->prepare("SELECT id FROM users WHERE id = ?", [$attr]), function($val) {
+								return \Amp\pipe($val->rowCount(), function($count) {
+									if ($count == 0) {
+										return 1; // error, add identifier for message
+									}
+								});
+							});
+						}
+						$promises[] = \Amp\pipe($this->db->prepare("SELECT id FROM users WHERE name = ?", [$attr]), function($val) use (&$attr) {
+							return \Amp\pipe($val->fetchObject(), function($row) use (&$attr) {
+								if ($row) {
+									$attr = $row->id;
+								} else {
+									return 1; // error, add identifier for message
+								}
+							});
+						});
+						break;
+					case self::INT:
+						if (($attr = filter_var($attr, FILTER_VALIDATE_INT)) === false) {
+							return null;
+						}
+					case self::ENUM:
+						foreach ($info["values"] as $val) {
+							if ($attr == $val["name"]) {
+								break 2;
+							}
+						}
+						return null;
+				}
+
+				$add[] = &$id;
+				$add[] = $field;
+				$add[] = &$attr;
+				unset($attr);
+			}
+		}
+
+		if ($errors = yield \Amp\filter($promises)) {
+			return null;
+		}
 
-		return ["id" => $info->insertId];
+		$conn = yield $this->db->getConnection();
+		$conn->query("START TRANSACTION");
+		
+		if ($id = $routed["id"] ?? 0) {
+			$info = yield $conn->prepare("UPDATE bugs SET title = :title, data = :data WHERE id = :id", ["id" => $id, "title" => $title, "data" => $data]);
+			if (!$info->affectedRows) {
+				return null;
+			}
+			yield $conn->prepare("DELETE FROM bugattrs WHERE bug = :id", ["id" => $id]);
+		} else {
+			$session = $req->getLocalVar(RequestKeys::SESSION);
+			$uid = $session->get(SessionKeys::LOGIN);
+			$info = yield $conn->prepare("INSERT INTO bugs (title, data, submitter) VALUES (:title, :data, :submitter)", ["title" => $title, "data" => $data, "submitter" => $uid]);
+			$id = $info->insertId;
+		}
+
+		if ($add) {
+			yield $conn->prepare("INSERT INTO bugattrs (bug, attribute, value) VALUES (?, ?, ?)" . str_repeat(", (?, ?, ?)", count($add) / 3 - 1), $add);
+		}
+		
+		$conn->query("COMMIT");
```

Comment:
absolutely. the INSERTS do nothing, until they are commited (when using transactions)